File: submissions/0035-search-insert-position/solution.java
class Solution {
    public int searchInsert(int[] nums, int target) {
        return search(nums, target, 0, nums.length-1);
    }
    
    private int search(int[] nums, int target, int left, int right) {
        
        if(left>right) {
            return left;
        }
        int currentIdx = left + (right-left)/2;
        int current = nums[currentIdx];
        if(target==current) {
            return currentIdx;
        }
        if(target < current) {
            return search(nums, target, left, currentIdx-1);
        } else {
            return search(nums, target, currentIdx+1, right);
        }
    }
}

File: submissions/0035-search-insert-position/solution.py
class Solution:
    def searchInsert(self, nums: List[int], target: int) -> int:
        idx = len(nums)
        lo = 0
        hi = len(nums)

        # search idx of upper bound
        while lo < hi:
            mid = (lo+hi) // 2
            val = nums[mid]

            if val == target:
                return mid
            else:
                if val > target:
                    idx = mid
                    hi = mid
                else:
                    lo = mid + 1
        
        return idx
        
        
